# HG changeset patch
# User Rik <rik@octave.org>
# Date 1504134344 25200
#      Wed Aug 30 16:05:44 2017 -0700
# Node ID 529272d1c271d50f103ae64bedb089b55d91ca30
# Parent  976a7a3502745800b94f6ec6a6d84d094fdaae1c
Complete openvar implementation (bug #51899).

* NEWS: Announce Variable Editor.  Add openvar to list of new functions.

* octave-link.cc (Fopenvar): Add "#include builtin-defun-decls.h" for access
to Octave functions.  Sort #includes.  Use is_string() to validate NAME input.
Call Fisguirunning and issue a warning if it is not.  Add BIST tests

* scripts/help/__unimplemented__.m: Remove openvar from list.

diff --git a/NEWS b/NEWS
--- a/NEWS
+++ b/NEWS
@@ -1,11 +1,17 @@
 Summary of important user-visible changes for version 4.4 (yyyy-mm-dd):
 ----------------------------------------------------------------------
 
+ ** A graphical Variable Editor has been added to the GUI interface.
+    It uses a spreadsheet-like interface for quick, intuitive editing
+    of variables.  The Variable Editor is launched by double-clicking
+    on a variable name in the Workspace Window or by typing
+    "openvar VARIABLE_NAME" in the Command Window.
+
  ** On systems with 64-bit pointers, --enable-64 is now the default and
     Octave always uses 64-bit indexing.  However, if the configure
     script determines that the BLAS library uses 32-bit integers, then
     operations using the following libraries are limited to arrays with
     dimensions that are smaller than 2^31 elements:
 
       BLAS  LAPACK  QRUPDATE  SuiteSparse  ARPACK
 
@@ -37,16 +43,17 @@ Summary of important user-visible change
 
  ** Other new functions added in 4.4:
 
       corrcoef
       getframe
       gsvd
       hgtransform
       humps
+      openvar
       repelem
 
  ** Deprecated functions.
 
     The following functions have been deprecated in Octave 4.4 and will
     be removed from Octave 4.8 (or whatever version is the second major
     release after 4.4):
 
diff --git a/libinterp/corefcn/octave-link.cc b/libinterp/corefcn/octave-link.cc
--- a/libinterp/corefcn/octave-link.cc
+++ b/libinterp/corefcn/octave-link.cc
@@ -21,23 +21,25 @@ along with Octave; see the file COPYING.
 <http://www.gnu.org/licenses/>.
 
 */
 
 #if defined (HAVE_CONFIG_H)
 #  include "config.h"
 #endif
 
+#include "builtin-defun-decls.h"
 #include "cmd-edit.h"
 #include "defun.h"
+#include "interpreter.h"
 #include "interpreter-private.h"
-#include "interpreter.h"
+#include "octave-link.h"
 #include "oct-env.h"
 #include "oct-mutex.h"
-#include "octave-link.h"
+#include "ovl.h"
 #include "pager.h"
 
 static int
 octave_readline_hook (void)
 {
   octave_link::entered_readline_hook ();
   octave_link::generate_events ();
   octave_link::process_events ();
@@ -383,32 +385,38 @@ Undocumented internal function.
 }
 
 DEFUN (openvar, args, ,
        doc: /* -*- texinfo -*-
 @deftypefn {} {} openvar (@var{name})
 Open the variable @var{name} in the graphical Variable Editor.
 @end deftypefn */)
 {
-  octave_value retval;
-
-  if (args.length () == 1)
-    {
-      std::string name = args (0).string_value ();
-      if (! error_state)
-        octave_link::openvar (name);
-      else
-        error ("invalid arguments");
-    }
-  else
+  if (args.length () != 1)
     print_usage ();
 
-  return retval;
+  if (! args(0).is_string ())
+    error ("openvar: NAME must be a string"); 
+
+  std::string name = args(0).string_value ();
+
+  if (! (Fisguirunning ())(0).is_true ())
+    warning ("openvar: GUI is not running, can't start Variable Editor"); 
+  else
+    octave_link::openvar (name);
+
+  return ovl ();
 }
 
+/*
+%!error openvar ()
+%!error openvar ("a", "b")
+%!error <NAME must be a string> openvar (1:10)
+*/
+
 DEFUN (__octave_link_show_doc__, args, ,
        doc: /* -*- texinfo -*-
 @deftypefn {} {} __octave_link_show_doc__ (@var{filename})
 Undocumented internal function.
 @end deftypefn */)
 {
   std::string file;
 
diff --git a/scripts/help/__unimplemented__.m b/scripts/help/__unimplemented__.m
--- a/scripts/help/__unimplemented__.m
+++ b/scripts/help/__unimplemented__.m
@@ -749,17 +749,16 @@ function list = missing_functions ()
   "nearestNeighbor",
   "netcdf",
   "noanimate",
   "notebook",
   "notify",
   "odextend",
   "openfig",
   "opengl",
-  "openvar",
   "ordeig",
   "ordqz",
   "outerjoin",
   "parseSoapResponse",
   "partition",
   "pathtool",
   "pcode",
   "pdepe",
